dhall/src/Dhall/Core.hs

```diff
@@ -1754,6 +1759,11 @@ isNormalizedWith :: (Eq s, Eq a) => Normalizer a -> Expr s a -> Bool
 isNormalizedWith ctx e = e == normalizeWith (Just (ReifiedNormalizer ctx)) e
 
 -- | Quickly check if an expression is in normal form
+--
+-- Given a well-typed expression @e@, @'isNormalized' e@ is equivalent to
+-- @e == 'normalize' e@.
+--
+-- Given an ill-typed expression, 'isNormalized' may return 'True' or 'False'.
```

Comment:
@Gabriel439 What do you think about this "contract"?

Otherwise I'd propose returning a `Maybe Bool` that is `Nothing` when normalization should fail.

Or maybe a custom type like `data Normalized = NormalForm | NonNormalForm | IllTyped` or so?

Reply:
I've played around with [a `Maybe Bool`-version of `isNormalized`](https://github.com/sjakobi/dhall-haskell/tree/isNormalized-Maybe) a bit. It seems rather tricky to nail down the consistency with `normalize` though. I think this should wait until `normalize` itself is more principled, possibly after the NbE changes.